### PR TITLE
romio/adio: add support for DDN's Infinite Memory Engine (IME)

### PR DESCRIPTION
**Pull request by sydidelot**
Hi,
This PR adds the support for DDN's Infinite Memory Engine (IME) to romio.
@roblatham00, thank you in advance for taking the time to review my patch.

Signed-off-by: Sylvain Didelot <sdidelot@ddn.com>

### CODE REVIEW

File: src/mpi/romio/adio/ad_ime/ad_ime_delete.c [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    static char myname[] = "ADIOI_IME_DELETE";
+
+    char *ime_filename = ADIOI_IME_Add_prefix(filename);
+    ret = ime_native_unlink(ime_filename);
+    ADIOI_Free(ime_filename);
+    if (ret)
+        *error_code = MPIO_Err_create_code(MPI_SUCCESS,
+                 MPIR_ERR_RECOVERABLE,
+                 myname, __LINE__,
+                 ADIOI_IME_error_convert(ret),
```
Comment (roblatham00) [inline src/mpi/romio/adio/ad_ime/ad_ime_delete.c]: Where is `ADIOI_IME_error_convert` defined?  Did this patch forget a file?

**Comment by roblatham00** [comment]
Another comment: where do users get the DDN IME library/header?  I mocked up the ime_native routines far enough to get ROMIO to compile and link, and the ROMIO "runtests" seems to work well enough.    When we document the new feature, though, I'd like to be able to tell users 'download from xyz" or "contact your DDN rep" or whatever the process is.

File: src/mpi/romio/adio/ad_ime/ad_ime_delete.c (hunk 1 above)
Comment (sydidelot) [reply src/mpi/romio/adio/ad_ime/ad_ime_delete.c]: Thanks for reviewing the patches! 
I've figured out that few changes were missing in the patch (I've probably forgotten to add the changes to the git commit). Thanks for noticing this! 
I'm going to submit a new version of the patch in a few minutes.

**Comment by sydidelot** [comment]
Rob,  `DEFAULT_IME_PREFIX_NO_FWD_SLASH` and `IME_FILE_PREFIX_LEN_NO_FWD_SLASH` are both defined in `ime_native.h`. If you are interested, you can access a copy of the header file for IME native API here: https://github.com/sydidelot/imenative/blob/master/ime_native.h
If you prefer, I can move the definition of those macros directly into ROMIO.

> Another comment: where do users get the DDN IME library/header? I mocked up the ime_native routines far enough to get ROMIO to compile and link, and the ROMIO "runtests" seems to work well enough. When we document the new feature, though, I'd like to be able to tell users 'download from xyz" or "contact your DDN rep" or whatever the process is.

IME is a proprietary software with closed sources. It is shipped in RPM packages with pre-compiled binaries. To compile a code that makes use of IME's native API, the user needs to include `ime_native.h` and link against `libim_client.so`. For the documentation about the new feature, I'm going to ask DDN what is the procedure. I'll get back to you with this information.

**Comment by roblatham00** [comment]
Great.  Your `ime_native.h` looks a lot like my mock-up.  

I guess ROMIO is IME-Native's only client, but the idea of a prefix and a prefix length... that's a ROMIO specific demand.  Stick it in `src/mpi/romio/adio/ad_ime/ad_ime.h` , please. 

Users would get the "IME Native"  library the same way they get the Panasas SDK.  Perfectly fine, just want to make sure I document the new feature correctly.

**Comment by sydidelot** [comment]
@roblatham00, as requested I moved the definition of `DEFAULT_IME_PREFIX_NO_FWD_SLASH` and `IME_FILE_PREFIX_LEN_NO_FWD_SLASH` into `src/mpi/romio/adio/ad_ime/ad_ime.h`. Also, I decided to rename those definitions to `ADIOI_IME_PREFIX` and `ADIOI_IME_PREFIX_LEN`, which is more meaningful ;)

**Comment by roblatham00** [comment]
Thanks.  Looks like something ate one last bit of feedback I had for you:

`ADIOI_IME_error_convert` is used but not defined anywhere.

**Comment by sydidelot** [comment]
Sorry about this, I thought I replied. 
`ADIOI_IME_error_convert` was a mistake and should have been removed: the first patch was missing some changes, certainly because I forgot to add them in the git commit.
The latest patch currently on Github must be OK.

**Comment by roblatham00** [comment]
OK, looks good.  I've signed off on this change and pushed to master.  Thanks for your contribution and for responding to my feedback so promptly.

**Comment by sydidelot** [comment]
Thank you very much for your time reviewing my patch, @roblatham00

**Pull request closed without merging**